keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeDelayed AllocationMode = "Delayed"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It's necessary to support
+	// deallocation when the class gets deleted before a claim.
+	DriverName string
+
+	// Scheduling contains information that is only relevant while the
+	// scheduler and the resource driver are in the process of selecting a
+	// node for a Pod and the allocation mode is AllocationModeDelayed. The
+	// resource driver should unset this when it has successfully allocated
+	// the resource.
+	Scheduling SchedulingStatus
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// Deallocate may be set to true to request freeing a resource as soon
+	// as it is unused. The scheduler uses this when it finds that freeing
+	// the resource and reallocating it elsewhere might unblock a pod.
+	//
+	// The resource driver checks this fields and resets it to false
+	// together with clearing the Allocation field. It also sets it
+	// to false when the resource is not allocated.
+	Deallocate bool
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the resource.  Usually those are Pods, but any other object that
+	// currently exists is also possible.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
+	// in an atomic ResourceClaim update because there might be multiple
+	// schedulers working on different Pods that compete for access to the
+	// same ResourceClaim.
+	//
+	// kubelet will check this before allowing a Pod to run because a
+	// scheduler might have missed that step, for example because it
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	ReservedFor []metav1.OwnerReference
+
+	// UsedOnNodes is a list of nodes where the ResourceClaim is or is
+	// going to be used. This must be set by the scheduler after scheduling
+	// a Pod onto a node.
+	//
+	// List/watch requests for ResourceClaims can filter on this field
+	// using a "status.usedOnNodes.<entry>=1" fieldSelector. kubelet uses
+	// this to limit which ResourceClaims it receives from the apiserver.
+	UsedOnNodes []string
```

Comment:
Is this bounded?  E.g. could I, theoretically, use a resource on 5000 nodes at the same time?  We can't have unbounded lists.

Reply:
Same as for `ReservedFor`: we could impose some artificial limit.

This field was added to address the scalability concern that @derekwaynecarr raised about all kubelets having to receive all ResourceClaims. With this field, they can set up an informer and do server-side filtering.

But perhaps we don't need it? Instead of using an informer, kubelet can also do a real GET if (and only if) it needs to. This will lead to additional latency during pod startup and increase the number of requests per pod startup, but it would still scale linearly with the number of pods.

Reply:
We have to balance - id we think that a resource could be used on O(1000) nodes at once, this simply won't scale.  We have to limit this to like 100 or 500 or else find a different pattern.

Reply:
Can we leave introducing bounds out of the KEP for now? We don't know yet which kind of resources will support an unlimited number of users, so whatever number we choose now would be speculative and not driven by any real measurements.

But let's consider whether UsedOnNodes is needed at all (see my comment above). Shall we start simple and remove it from the proposal?

Reply:
I have removed it because:
- it's simpler
- we don't know whether informers are needed in kubelet
- last but not least, keeping it up-to-date would need write permission for kubelet (removed now) or additional code in the resource controller